### PR TITLE
[SPARK-16628][SQL] Translate file-based relation schema when file schema is inconsistent with catalog schema

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

We will convert Metastore Orc tables (represented by `MetastoreRelation`) to datasource tables (represented by `HadoopFsRelation`) if `spark.sql.hive.convertMetastoreOrc` is enabled for better performance.

However, due to a Hive issue, an Orc table created by Hive does not store column name correctly in the Orc files. For these Orc tables, the converted relation has wrong schema.

To fix this, we assume the metastore schema `dataSchema` can match to `physicalSchema` by each column disregarding the column names. If not, we throw an exception that suggests users to disable the conversion of Hive Orc tables.

## How was this patch tested?

Jenkins tests.

### CODE REVIEW

**Comment by viirya** [comment]
cc @cloud-fan @yhuai @liancheng @rxin Please review this change. Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SchemaMapping.scala [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+package org.apache.spark.sql.execution
+
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression}
+import org.apache.spark.sql.types.{StructField, StructType}
+
+/**
+ * An interface for mapping two different schemas. For the relations that have are backed by files,
+ * the inferred schema from the files might be different with the schema stored in the catalog. In
+ * such case, the interface helps mapping inconsistent schemas.
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SchemaMapping.scala]: Can you put the detailed description of this mapping in the doc here? thanks.

Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SchemaMapping.scala]: I've added more detailed document. Please take a look. Thanks.

**Comment by viirya** [comment]
cc @yhuai @liancheng @rxin

**Comment by cloud-fan** [comment]
@rxin @yhuai , is it worth adding this complex logic to work around this hive bug? This feature(convert Metastore Orc tables to data source table) is not a major feature and the broken case seems a corner case. I think it's ok to disable this feature when we meet this hive bug, what do you think?

**Comment by viirya** [comment]
@cloud-fan I've submitted a PR #14282 previously that disables the conversion if the schema is inconsistent.

**Comment by viirya** [comment]
@cloud-fan So do we have decision on this? A simpler approach to disable the conversion if the schema is inconsistent, or a complex one to work around this Hive bug?

**Comment by viirya** [comment]
ping @cloud-fan Can you have a decision about this? Are we going to have complex logic for this issue? Or just disable it?

**Comment by cloud-fan** [comment]
I noticed that `spark.sql.hive.convertMetastoreOrc` is false by default, so this bug is not so critical.

And we may stop inferring file schema when reading ORC tables, for performance reasons(#16980), then we don't have the chance to detect the mismatch and disable this feature.

We may remove this feature entirely, or thinking harder to come up with a better fix. cc @yhuai

**Comment by yhuai** [comment]
@cloud-fan actually, this conversion was disabled because of this bug. 

btw, pr that @cloud-fan mentioned is https://github.com/apache/spark/pull/14690. 

I think it is better to hold this change before we finish https://github.com/apache/spark/pull/14690.

**Comment by dongjoon-hyun** [comment]
Hi, @viirya , @cloud-fan , @yhuai .
#14690 and the parent issue SPARK-17861 seems to be resolved?
What about restarting this issue?

**Comment by viirya** [comment]
As we replace schema inferring with metastore schema completely by #14690 for converted Hive tables, we may not have the chance to detect the mismatch between Orc file's physical schema and metastore schema.

**Comment by dongjoon-hyun** [comment]
It seems to exist some cases like the following (on the current master).

On hive,
```sql
CREATE TABLE t1 (a string) PARTITIONED BY (b string) STORED AS ORC;
INSERT INTO TABLE t1 PARTITION (b='01') VALUES ('1');
```

On `spark-shell`,
```scala
scala> sql("select * from t1").show
+---+---+
|  a|  b|
+---+---+
|  1| 01|
+---+---+

scala> sql("set spark.sql.hive.convertMetastoreOrc=true").show
+--------------------+-----+
|                 key|value|
+--------------------+-----+
|spark.sql.hive.co...| true|
+--------------------+-----+

scala> sql("select * from t1").show
16/12/07 05:17:19 ERROR Executor: Exception in task 0.0 in stage 2.0 (TID 2)
java.lang.IllegalArgumentException: Field "a" does not exist.
```

**Comment by viirya** [comment]
@dongjoon-hyun yeah, I see. Because we directly use metastore schema of converted Orc table, when the physical schema in Orc file and metastore schema mismatch, this issue happens.

**Comment by viirya** [comment]
We have two options. First one is to map metastore schema to physical Orc schema like this. But we don't infer physical schema of Orc file now. I will update this to have this mapping in OrcFileFormat.

Another one is like #14282. But as we don't infer schema from Orc file now, we can't disable the conversion when the mismatch is detected. One possible is to throw exception in OrcFileFormat when detecting the mismatch before reading and show message to ask user to disable `spark.sql.hive.convertMetastoreOrc`.

@cloud-fan @yhuai @dongjoon-hyun What do you think?

**Comment by viirya** [comment]
@cloud-fan @yhuai @dongjoon-hyun I've updated this as:

* Assume metastore schema matches with physical Orc schema by column, disregarding column names. 

* Mapping required schema to columns in physical Orc schema.

* If the length or data types of metastore schema and physical schema is not matched, throw an exception suggesting users to disable `spark.sql.hive.convertMetastoreOrc`.

Please let me know what you think about this approach. Thanks.

**Comment by viirya** [comment]
retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala [hunk 2]
```diff
@@ -303,6 +330,10 @@ private[orc] object OrcRelation extends HiveInspectors {
     maybeStructOI.map(unwrap).getOrElse(Iterator.empty)
   }
 
+  def isMismatchSchema(physicalSchema: StructType, requestedSchema: StructType): Boolean = {
+    requestedSchema.forall(a => physicalSchema.getFieldIndex(a.name).isEmpty)
```
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala]: Oh, never mind. It's really about checking whether all requested columns are not matched.

**Comment by dongjoon-hyun** [comment]
It looks good to me. Thank you for updating, @viirya !

**Comment by viirya** [comment]
ping @cloud-fan @yhuai May you take a look? Thanks.

**Pull request closed without merging**